File: Yixuan/17_01_17/total_occurrence_of_target.cpp
class Solution {
public:
    /**
     * @param A an integer array sorted in ascending order
     * @param target an integer
     * @return an integer
     * @link: http://www.lintcode.com/en/problem/total-occurrence-of-target/
     * @author: Egbert Li
     */
    int totalOccurrence(vector<int>& A, int target) {
       int length = A.size();
       if (length == 0) {
           return 0;
       }
       if (target > A[length - 1] || target < A[0]) {
           return 0;
       }
       int left = 0;
       int right = 0;
       int start = 0;
       int end = length - 1;
       // get lower bound, if A[start] != A[end] != target, then return 0
       while (start + 1 < end) {
           int mid = start + ((end - start) >> 1);
           if (target > A[mid]) {
               start = mid;
           } else {
                end = mid;
           }
       }
       if (target == A[start]) {
           left = start;
       } else if (target == A[end]) {
           left = end;
       } else {
           return 0;
       }

       // get upper bound
       start = 0;
       end = length - 1;
       while (start + 1 < end) {
           int mid = start + ((end - start) >> 1);
           if (target < A[mid]) {
               end = mid;
           } else {
               start = mid;
           }
       }
       if (target == A[end]) {
           right = end;
       } else {
           right = start;
       }
       return  right - left + 1;
    }
};

File: Boxuan/16_01_17/total_occurrence_of_target.py
# Python
class Solution:
    # @param {int[]} A an integer array sorted in ascending order
    # @param {int} target an integer
    # @return {int} an integer
    # @link: http://www.lintcode.com/en/problem/total-occurrence-of-target/
    # @author: Egbert Li
    def totalOccurrence(self, A, target):
        if not A:
            return 0
        
        left, right = 0, len(A) - 1

#search the first left index of target
        while left + 1 < right:
            mid = left + (right - left) / 2
            if A[mid] < target:
                left = mid
            else:
                right = mid
        left_index = left if A[left] == target else right
        if left_index != target:
            return -1
#search the first right index of target

        left, right = 0, len(A) - 1
        while left + 1 < right:
            mid = left + (right - left) / 2
            if A[mid] > target:
                right = mid
            else:
                left = mid
        right_index = right if A[right] == target else left

        return abs(right_index - left_index) + 1